enhancements/two-node-fencing/tnf.md

```diff
@@ -0,0 +1,767 @@
+
+RHEL-HA has no real understanding of the resources (IP addresses, file systems, databases, even virtual machines) it manages.
+It relies on resource agents to understand how to check the state of a resource, as well as start and stop them to achieve the desired target state.
+How a given agent uses these actions, and associated states, to model the resource is opaque to the cluster and depends on the needs of the underlying resource.
+
+Resource agents must conform to one of a variety of standards, including systemd, SYS-V, and OCF.
+The latter is the most powerful, adding the concept of promotion, and demotion.
+More information on creating OCF agents can be found in the upstream [developer guide](https://github.com/ClusterLabs/resource-agents/blob/main/doc/dev-guides/ra-dev-guide.asc).
+
+Tools for extracting support information (must-gather tarballs) will be updated to gather relevant logs for triaging issues.
+
+As part of the fencing setup, the cri-o and kubelet services will still be owned by systemd when running under pacemaker. The main difference is that the resource agent will be responsible for signaling systemd to change their active states.
+The etcd containers are different in this respect since they will be restarted using Podman, but this will be running as root, as it was under CEO.
+
+#### Platform None vs. Baremetal
+One of the major design questions of two-node OpenShift is whether to target support for `platform: none` or `platform: baremetal`. The advantage of selecting `platform: baremetal` is that we can leverage the benefits of deploying an ingress-VIP out of the box using keepalived and haproxy. After some discussion with the metal networking team, it is expected that this might work without modifications as long as pacemaker fencing doesn't remove nodes from the node list so that both keepalived instances are always peers. Furthermore, it was noted that this might be solved more simply without keepalived at all by using the ipaddr2 resource agent for pacemaker to run the `ip addr add` and `ip addr remove` commands for the VIP.
+The bottom line is that it will take some engineering effort to modify the out-of-the-box in-cluster networking feature for two-node OpenShift.
+
+Outside of potentially reusing the networking bits of `platform: baremetal`, we discussed potentially reusing its API for collecting BMC credentials for fencing. In this approach, we'd use the `platform: baremetal` BMC entries would be loaded into BareMetalHost CRDs and we'd extend BMO to initialize pacemaker instead of a new operator. After a discussion with the Baremetal Platform team, we were advised against using the Baremetal Operator as an inventory. Its purpose/scope is provisioning nodes.
```

Comment:
This is not at all the reason.
The reason is that HA of the baremetal platform components is very much dependent on a working read-write control plane.
Fencing is absolutely part of the purpose/scope of BMO. However, users have already noted that for a 3-node (compact) cluster failover is too slow, and for a 2-node cluster it provably cannot work at all.

Reply:
I think I misinterpreted the intention of [this comment](https://redhat-internal.slack.com/archives/CFP6ST0A3/p1733384274212569?thread_ts=1733356326.224649&cid=CFP6ST0A3).
This section should explain that the existing BMO-based fencing is insufficient to meet our requirements and that integrating with BMO to reuse the existing APIs that BMO exposes would add a dependency on BMO to behave like a baremetal inventory - which is not its purpose.

Reply:
The way I see it, acting as a baremetal inventory _is_ part of the purpose of BMO.
It just won't work for this use case.

Reply:
Understood! Thanks for taking the time to clarify! :)